File: src/devices/gopro/protocol.cpp
#include "devices/gopro/protocol.h"

namespace gopro {

Packet buildSetShutter(bool enabled) {
  Packet packet;
  packet.bytes[0] = 0x03;
  packet.bytes[1] = kSetShutterCommand;
  packet.bytes[2] = 0x01;
  packet.bytes[3] = enabled ? 0x01 : 0x00;
  packet.len = 4;
  return packet;
}

Packet buildSleep() {
  Packet packet;
  packet.bytes[0] = 0x01;
  packet.bytes[1] = kSleepCommand;
  packet.len = 2;
  return packet;
}

Packet buildSetPairingState() {
  Packet packet;
  packet.bytes[0] = 0x03;
  packet.bytes[1] = kSetPairingStateCommand;
  packet.bytes[2] = 0x01;
  packet.bytes[3] = 0x01;
  packet.len = 4;
  return packet;
}

Packet buildGetHardwareInfo() {
  Packet packet;
  packet.bytes[0] = 0x01;
  packet.bytes[1] = kGetHardwareInfoCommand;
  packet.len = 2;
  return packet;
}

Packet buildGetEncoding(bool twoByteIds) {
  Packet packet;
  if (twoByteIds) {
    packet.bytes[0] = 0x03;
    packet.bytes[1] = kGetStatusValues2Byte;
    packet.bytes[2] = 0x00;
    packet.bytes[3] = kEncodingStatus;
    packet.len = 4;
  } else {
    packet.bytes[0] = 0x02;
    packet.bytes[1] = kGetStatusValues;
    packet.bytes[2] = kEncodingStatus;
    packet.len = 3;
  }
  return packet;
}

Packet buildRegisterEncoding(bool twoByteIds) {
  Packet packet;
  if (twoByteIds) {
    packet.bytes[0] = 0x03;
    packet.bytes[1] = kRegisterStatusUpdates2Byte;
    packet.bytes[2] = 0x00;
    packet.bytes[3] = kEncodingStatus;
    packet.len = 4;
  } else {
    packet.bytes[0] = 0x02;
    packet.bytes[1] = kRegisterStatusUpdates;
    packet.bytes[2] = kEncodingStatus;
    packet.len = 3;
  }
  return packet;
}

void PacketAccumulator::reset() {
  expected_ = 0;
  length_ = 0;
  continuation_ = 0;
}

bool PacketAccumulator::feed(const uint8_t* data, size_t len,
                             Message& message) {
  message.len = 0;
  if (data == nullptr || len == 0) return false;
  size_t offset = 0;
  if ((data[0] & 0x80) != 0) {
    if (expected_ == 0 || (data[0] & 0x0f) != continuation_) {
      reset();
      return false;
    }
    continuation_ = static_cast<uint8_t>((continuation_ + 1) & 0x0f);
    offset = 1;
  } else {
    reset();
    const uint8_t headerType = static_cast<uint8_t>((data[0] >> 5) & 0x03);
    if (headerType == 0) {
      expected_ = data[0] & 0x1f;
      offset = 1;
    } else if (headerType == 1 && len >= 2) {
      expected_ = (static_cast<size_t>(data[0] & 0x1f) << 8) | data[1];
      offset = 2;
    } else if (headerType == 2 && len >= 3) {
      expected_ = (static_cast<size_t>(data[1]) << 8) | data[2];
      offset = 3;
    } else {
      reset();
      return false;
    }
    if (expected_ == 0 || expected_ > sizeof(bytes_)) {
      reset();
      return false;
    }
  }
  const size_t incoming = len - offset;
  if (length_ + incoming > expected_ || length_ + incoming > sizeof(bytes_)) {
    reset();
    return false;
  }
  for (size_t i = 0; i < incoming; ++i) bytes_[length_ + i] = data[offset + i];
  length_ += incoming;
  if (length_ != expected_) return false;
  message.len = length_;
  for (size_t i = 0; i < length_; ++i) message.bytes[i] = bytes_[i];
  reset();
  return true;
}

Response parseCommandPayload(const uint8_t* data, size_t len) {
  Response response;
  if (data == nullptr || len < 2) return response;
  response.valid = true;
  response.command = data[0];
  response.status = data[1];
  return response;
}

Response parseResponse(const uint8_t* data, size_t len) {
  Response response;
  if (data == nullptr || len < 3 || data[0] < 2 || data[0] + 1 > len) {
    return response;
  }
  return parseCommandPayload(data + 1, data[0]);
}

StatusResponse parseStatusPayload(const uint8_t* data, size_t len) {
  StatusResponse response;
  if (data == nullptr || len < 2) return response;
  const bool twoByteIds = data[0] == kGetStatusValues2Byte ||
                          data[0] == kRegisterStatusUpdates2Byte ||
                          data[0] == kNotifyStatusUpdate2Byte;
  if (!twoByteIds && data[0] != kGetStatusValues &&
      data[0] != kRegisterStatusUpdates &&
      data[0] != kNotifyStatusUpdate) {
    return response;
  }
  response.valid = true;
  response.operation = data[0];
  response.status = data[1];
  size_t position = 2;
  while (position < len) {
    const size_t headerLength = twoByteIds ? 3 : 2;
    if (position + headerLength > len) return StatusResponse{};
    const uint16_t id = twoByteIds
                            ? static_cast<uint16_t>(data[position] << 8) |
                                  data[position + 1]
                            : data[position];
    const uint8_t valueLength = data[position + headerLength - 1];
    position += headerLength;
    if (position + valueLength > len) return StatusResponse{};
    if (id == kEncodingStatus) {
      if (valueLength != 1 || data[position] > 1) return StatusResponse{};
      response.hasEncoding = true;
      response.encoding = data[position] == 1;
    }
    position += valueLength;
  }
  return response;
}

}  // namespace gopro

File: tools/gopro_lab/protocol.py
from __future__ import annotations

from dataclasses import dataclass


SERVICE_UUID = "0000fea6-0000-1000-8000-00805f9b34fb"
COMMAND_UUID = "b5f90072-aa8d-11e3-9046-0002a5d5c51b"
COMMAND_RESPONSE_UUID = "b5f90073-aa8d-11e3-9046-0002a5d5c51b"
QUERY_UUID = "b5f90076-aa8d-11e3-9046-0002a5d5c51b"
QUERY_RESPONSE_UUID = "b5f90077-aa8d-11e3-9046-0002a5d5c51b"

SET_SHUTTER = 0x01
SET_PAIRING_STATE = 0x17
GET_HARDWARE_INFO = 0x3C
GET_STATUS_VALUES = 0x13
REGISTER_STATUS_UPDATES = 0x53
NOTIFY_STATUS_UPDATE = 0x93
GET_STATUS_VALUES_2BYTE = 0x16
REGISTER_STATUS_UPDATES_2BYTE = 0x56
NOTIFY_STATUS_UPDATE_2BYTE = 0x96
ENCODING_STATUS = 0x0A
SUCCESS = 0x00


def packet(payload: bytes) -> bytes:
    """Wrap one small Open GoPro payload in a General packet header."""
    if not payload or len(payload) > 20:
        raise ValueError("the lab only emits non-empty payloads up to 20 bytes")
    return bytes((len(payload),)) + payload


def set_shutter(enabled: bool) -> bytes:
    return packet(bytes((SET_SHUTTER, 0x01, 0x01 if enabled else 0x00)))


def set_pairing_state() -> bytes:
    return packet(bytes((SET_PAIRING_STATE, 0x01, 0x01)))


def get_hardware_info() -> bytes:
    return packet(bytes((GET_HARDWARE_INFO,)))


def get_encoding(two_byte_ids: bool = False) -> bytes:
    if two_byte_ids:
        return packet(bytes((GET_STATUS_VALUES_2BYTE, 0x00, ENCODING_STATUS)))
    return packet(bytes((GET_STATUS_VALUES, ENCODING_STATUS)))


def register_encoding(two_byte_ids: bool = False) -> bytes:
    if two_byte_ids:
        return packet(
            bytes((REGISTER_STATUS_UPDATES_2BYTE, 0x00, ENCODING_STATUS))
        )
    return packet(bytes((REGISTER_STATUS_UPDATES, ENCODING_STATUS)))


@dataclass(frozen=True)
class CommandResponse:
    command: int
    status: int
    data: bytes


@dataclass(frozen=True)
class StatusResponse:
    operation: int
    status: int
    encoding: bool | None


class PacketAccumulator:
    """Reassemble General, Extended, and Continuation Open GoPro packets."""

    def __init__(self) -> None:
        self._expected = 0
        self._payload = bytearray()
        self._counter = 0

    def reset(self) -> None:
        self._expected = 0
        self._payload.clear()
        self._counter = 0

    def feed(self, data: bytes) -> bytes | None:
        if not data:
            return None
        first = data[0]
        if first & 0x80:
            if self._expected == 0 or (first & 0x0F) != self._counter:
                self.reset()
                return None
            self._counter = (self._counter + 1) & 0x0F
            self._payload.extend(data[1:])
        else:
            self.reset()
            header_type = (first >> 5) & 0x03
            if header_type == 0:
                self._expected = first & 0x1F
                header_length = 1
            elif header_type == 1:
                if len(data) < 2:
                    return None
                self._expected = ((first & 0x1F) << 8) | data[1]
                header_length = 2
            elif header_type == 2:
                if len(data) < 3:
                    return None
                self._expected = (data[1] << 8) | data[2]
                header_length = 3
            else:
                return None
            if self._expected == 0:
                self.reset()
                return None
            self._payload.extend(data[header_length:])

        if len(self._payload) > self._expected:
            self.reset()
            return None
        if len(self._payload) != self._expected:
            return None
        payload = bytes(self._payload)
        self.reset()
        return payload


def decode_command_response(payload: bytes) -> CommandResponse | None:
    if len(payload) < 2:
        return None
    return CommandResponse(payload[0], payload[1], payload[2:])


def decode_status_response(payload: bytes) -> StatusResponse | None:
    if len(payload) < 2 or payload[0] not in (
        GET_STATUS_VALUES,
        REGISTER_STATUS_UPDATES,
        NOTIFY_STATUS_UPDATE,
        GET_STATUS_VALUES_2BYTE,
        REGISTER_STATUS_UPDATES_2BYTE,
        NOTIFY_STATUS_UPDATE_2BYTE,
    ):
        return None
    operation = payload[0]
    status = payload[1]
    position = 2
    encoding = None
    two_byte_ids = operation in (
        GET_STATUS_VALUES_2BYTE,
        REGISTER_STATUS_UPDATES_2BYTE,
        NOTIFY_STATUS_UPDATE_2BYTE,
    )
    while position < len(payload):
        header_length = 3 if two_byte_ids else 2
        if position + header_length > len(payload):
            return None
        if two_byte_ids:
            element_id = (payload[position] << 8) | payload[position + 1]
            element_length = payload[position + 2]
        else:
            element_id = payload[position]
            element_length = payload[position + 1]
        position += header_length
        if position + element_length > len(payload):
            return None
        value = payload[position : position + element_length]
        position += element_length
        if element_id == ENCODING_STATUS:
            if element_length != 1 or value[0] not in (0, 1):
                return None
            encoding = value[0] == 1
    return StatusResponse(operation, status, encoding)
